### PR TITLE
targets/ramips-mt76x8: add device TP-Link RE200 V4

### PR DESCRIPTION
**Pull request by misanthropos**
Signed-off-by: Richard Fröhning <misanthropos@gmx.de>

- [x] must be flashable from vendor firmware
  - [x] webinterface
  - [x] tftp
  - [ ] other: <specify>
- [x] must support upgrade mechanism
  - [x] must have working sysupgrade
    - [x] must keep/forget configuration (if applicable)
      *think `sysupgrade [-n]` or `firstboot`*
  - [x] must have working autoupdate
    *usually means: gluon profile name must match image name*
- [x] reset/wps/phone button must return device into config mode
- [x] primary mac should match address on device label (or packaging) (https://gluon.readthedocs.io/en/latest/dev/hardware.html#notes)
- wired network
  - [x] should support all network ports on the device
  - [x] must have correct port assignment (WAN/LAN)
    - On devices supplied via PoE, there is usually no explicit WAN/LAN labeling on the hardware.
      The PoE input should be the WAN port in this case.
- wifi (if applicable)
  - [ ] association with AP must be possible on all radios
  - [ ] association with 802.11s mesh must be working on all radios 
  - [ ] ap/mesh mode must work in parallel on all radios
- led mapping
  - power/sys led (_critical, because led definitions are setup on firstboot only_)
    - [x] lit while the device is on
    - [x] should display config mode blink sequence 
(https://gluon.readthedocs.io/en/latest/features/configmode.html)
  - radio leds
    - [x] should map to their respective radio
    - [x] should show activity
  - switchport leds
    - [x] should map to their respective port (or switch, if only one led present) 
    - [x] should show link state and activity
- outdoor devices only
  - [ ] added board name to `is_outdoor_device` function in `package/gluon-core/luasrc/usr/lib/lua/gluon/platform.lua`

### CODE REVIEW

**Comment by mweinelt** [comment]
Please fill out the device integration checklist and copy it into the top post.

https://github.com/freifunk-gluon/gluon/wiki/Device-Integration-checklist.

**Comment by blocktrron** [comment]
> 16:51 <Misanthropos> hexa-, thank you for the input. Added the Device Integration checklist. There is something odd however. The Power/Sys LED is off by default. In config-mode the LAN interface blinks. Changing trigger of power:green affects all leds.


Which revision is this referring to?

**Comment by misanthropos** [comment]
> Which revision is this referring to?

Both, V3 and V4 show the same behaviour.

**Comment by blocktrron** [comment]
In this case, this sounds like the GPIO assignment of the LEDs is off. Check the GPL tarball, maybe they are defined in there.

**Comment by AiyionPrime** [comment]
For the next one passing by:
@misanthropos is the author of https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=33b76d66d1afa9d888f32587b18d78a930593ca5.

**Comment by misanthropos** [comment]
I have found correct GPIO assignments which will find its way into OpenWRT with this PR:
https://github.com/openwrt/openwrt/pull/4954

**Comment by AiyionPrime** [comment]
@misanthropos in order to get this going again, might you start resolvin the conflicts?
maybe @adschm could then do another review.

**Comment by AiyionPrime** [comment]
tested backport patch is open for review
https://patchwork.ozlabs.org/project/openwrt/patch/20221108123102.666033-1-openwrt@aiyionpri.me/

**Comment by AiyionPrime** [comment]
currently blocked by #2698

**Comment by AiyionPrime** [comment]
@misanthropos This needs a rebase, as well as testing.
As your previous tests suggest, this revision might have trouble with at least one of its wifi interfaces and as such should be marked as broken, until it 's resolved.

File: docs/user/supported_devices.rst [hunk 1]
```diff
@@ -407,6 +407,7 @@ ramips-mt76x8
   - Archer C50 (v3)
   - Archer C50 (v4)
   - RE200 (v2)
+  - RE200 (v4)
```
Comment (AiyionPrime) [inline docs/user/supported_devices.rst:410]:
```suggestion
```

broken devices are not supported

**Review by AiyionPrime — CHANGES_REQUESTED** [review]

File: docs/user/supported_devices.rst (hunk 1 above)
Comment (misanthropos) [reply docs/user/supported_devices.rst:410]: I can close this PR and reopen it by the time its fixed?

Comment (AiyionPrime) [reply docs/user/supported_devices.rst:410]: You don't need to close this. You just need to remove it from the docs.
Once the issue is resolved you can open a PR that removes the `broken` tag and adds it as supported to the docs.

**Comment by blocktrron** [comment]
As @AiyionPrime told me, the 5 GHz interface does not work on channels above 64, right?

Can you provide a image of the EEPROM / caldata partition? The MT7612 can have channel allowance for controlling permitted channels with a single SKU programmed into it's EEPROM, i have never seen those but given it does seem to outright "not work", i assume this is the case here.

This might be fixable by extending mt76 to override this setting (which should be done anyways, given we use the wireless-regdb for this case).

**Comment by blocktrron** [comment]
@misanthropos 

Check which partition is labled eeprom with `cat /proc/mtd`, dump the content using `dd if=/dev/mtdX of=/tmp/eeprom.bin` and transfer this file to GitHub.

**Comment by misanthropos** [comment]
This is the result of cat /proc/mtd
```
root@re200v4:~# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00020000 00010000 "u-boot"
mtd1: 007a0000 00010000 "firmware"
mtd2: 0020f81b 00010000 "kernel"
mtd3: 005907e5 00010000 "rootfs"
mtd4: 00250000 00010000 "rootfs_data"
mtd5: 00030000 00010000 "config"
mtd6: 00010000 00010000 "radio"
```

**Comment by AiyionPrime** [comment]
@misanthropos dumps of `/dev/mtd/radio` (and `/dev/mtd/config` to be sure) should be sufficient.

**Comment by blocktrron** [comment]
@misanthropos can you try lowering the TX power? Start with txpower of 1 and work your way up until it does not work.

Looking at the code, "MAC error detected" seems to happen when beacons are stuck. This either indicates a bug in SW / HW or possibly noise on the channel, which might also be fedd back to the chip from itself.

My initial theory was that maybe the  MAC prohibits sending frames due to channels being excluded in EEPROM data. This is not the case, as EEPROM does not specify channel limits / regdom.

**Comment by misanthropos** [comment]
@blocktrron so here are my observations:

- Only txpower mode set to "auto" possible

´´´
iw client1 set txpower 1
Invalid parameter: 1
iw client1 set txpower fixed 10
command failed: Not supported (-122)
´´´
Same issue on re200v3.

Those are the defaults:
                  re200v3                                    re200v4
channel     44                                             36
power        11 dBm                                     10 dBm

**Comment by AiyionPrime** [comment]
@blocktrron I think this might be the commit, that introduced the behavior described above.
https://github.com/openwrt/mt76/commit/20d1fed838b958701ed5d030b663ba9227477ab1

As the issue was reportedly resolved:

> "MAC error detected" problem was fixed. Although channel 149-161 still doesn't work, it might be a hardware bug, so maybe we can close this issue.

_Originally posted by @DragonBluep in https://github.com/openwrt/mt76/issues/666#issuecomment-1166278390_

I would have expected @misanthropos reported problems are gone on a branch that contains the commit @nbd did two days earlier:

https://github.com/openwrt/mt76/commit/ba39ed3b44f1a0b3bd87c72452a261908afba754

That is the case for both OpenWrt's `master` and `openwrt-22.03` though.

**22.03** https://github.com/openwrt/mt76/commit/d70546462b7b51ebc2bcdd5c534fdf3465be62a4 referenced in https://github.com/openwrt/openwrt/blob/openwrt-22.03/package/kernel/mt76/Makefile#L12

That's from early september.
Looking at @misanthropos branch history (rebased 15 days ago) I'd assume they are included.

@blocktrron Any idea what I'm missing?

**Comment by misanthropos** [comment]
Needs reviewing of 5GHz - will reopen if it gets fixed in upstream.

**Comment by kpoman** [comment]
Hey guys, this is still failing on a fresh 23.05.0 installed on my RE200 V4... I just get mac errors. I tried different txpower levels with no luck.
`root@OpenWrt:~# logread | grep mt76x0e
Tue Nov 14 01:17:29 2023 kern.info kernel: [   12.597389] mt76x0e 0000:01:00.0: card - bus=0x1, slot = 0x0 irq=4
Tue Nov 14 01:17:29 2023 kern.info kernel: [   12.610061] mt76x0e 0000:01:00.0: ASIC revision: 76100002
Tue Nov 14 01:17:29 2023 kern.info kernel: [   12.700810] mt76x0e 0000:01:00.0: Firmware Version: 0.1.00
Tue Nov 14 01:17:29 2023 kern.info kernel: [   12.985556] mt76x0e 0000:01:00.0: EEPROM ver:01 fae:00
Tue Nov 14 01:17:58 2023 kern.err kernel: [   46.895026] mt76x0e 0000:01:00.0: MAC error detected
...
Tue Nov 14 01:30:50 2023 kern.err kernel: [  118.665109] mt76x0e 0000:01:00.0: MAC error detected
Tue Nov 14 01:30:51 2023 kern.err kernel: [  119.895105] mt76x0e 0000:01:00.0: MAC error detected
root@OpenWrt:~#`

**Comment by AiyionPrime** [comment]
Welcome @kpoman. This is the gluon repository.
We are currently not tracking this device as supported as we found the same flaws you just encountered.

Once the Issues are addressed in OpenWRT (if possible at all) we can continue this thread.
Until then I'd recommend OpenWRT's issue tracker for this bugreport.

**Pull request closed without merging**